### PR TITLE
Adjust opengraph tag to show scrapped image

### LINKED ISSUE #8273: adjust OpenGraph tags for questions/notes/wikis
### Please describe the problem (or idea)

adjust OpenGraph tags for questions/notes/wikis so that if no main image is available, openGraph lead image tags show a "scraped" lead image (see how we do this in card display, for example https://publiclab.org/tag/bucket-monitor shows a scraped lead image for https://publiclab.org/notes/kgradow1/07-03-2020/bucket-monitor-prototype)

will be making a PR for this soon

> What happened just before the problem occurred? Or what problem could this idea solve?

> What did you expect to see that you didn't?

### Please show us where to look

https://publiclab.org/ ...

### What's your PublicLab.org username?

> This can help us diagnose the issue: 

### Browser, version, and operating system

> Many bugs are related to these -- please help us track it down and reproduce what you're seeing!

****

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site. 

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### PR DESCRIPTION
**Pull request by Shulammite-Aso**
Fixes #8273 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts 📁
* [ ] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by Shulammite-Aso** [comment]
@jywarren @sagarpreet-chadha @cesswairimu please do check what i have here. I currently don't know how to confirm if this works or not.  Does this solve the problem?

Thanks!!!!

**Comment by Shulammite-Aso** [comment]
> Hi @Shulammite-Aso , I have replied on the other thread on how to test it using either `ngrok` or PL unstable, kindly check that

Thanks, seen that!

**Comment by jywarren** [comment]
Another possibility would be to push it to the `unstable` branch at https://github.com/publiclab/plots2/, and then it would build at http://unstable.publiclab.org. What do you think? You could also view source in the GitPod build once its running...

**Comment by jywarren** [comment]
Thank you @Shulammite-Aso !!

**Comment by jywarren** [comment]
This looks great. Testing it out now at:

* https://cards-dev.twitter.com/validator
* https://developers.facebook.com/tools/debug/

**Comment by jywarren** [comment]
In GitPod, I'm posting an image inline (not main image) and submitting the resulting posts to those test pages.

**Comment by jywarren** [comment]
Ah, due to the Redis issue fixed in https://github.com/publiclab/plots2/pull/8162, which is now resolved but this PR is not rebased over it, I'll try adding an inline image to an existing node and testing that, since we can't publish rn in GitPod due to Redis being missing.

**Comment by jywarren** [comment]
Hmm -- I'm not seeing the scraped image -- although I had to include a remote image to make it work. 

```
<meta property="og:title" content="Blog Post" />
--
  | <meta property="og:site_name" content="Public Lab" />
  | <meta property="og:url" content="https://3000-c6615f74-6744-49fd-be5d-597f27808a11.ws-us02.gitpod.io/n/8" />
  | <meta property="og:description" content="Blog post body
  |  
  |  
  | " />
  | <meta property="og:type" content="article" />
  | <meta property="fb:app_id" content="644692375883090" />
  | <meta property="og:image" content="https://3000-c6615f74-6744-49fd-be5d-597f27808a11.ws-us02.gitpod.io" />
  | <meta name="twitter:image" content="https://3000-c6615f74-6744-49fd-be5d-597f27808a11.ws-us02.gitpod.io">
```

Are you pretty sure it's returning the scraped image from a situation like this? Maybe we need to rebase to get the full ability to post new images inline in new posts?

https://github.com/publiclab/plots2/blob/27dbd9471f2029c2542086e98e330aea4380f995/app/models/node.rb#L350-L355

****

image

Facebook also seems not to have gotten it; https://developers.facebook.com/tools/debug/?q=https%3A%2F%2F3000-c6615f74-6744-49fd-be5d-597f27808a11.ws-us02.gitpod.io%2Fnotes%2Fadmin%2F08-18-2020%2Fblog-post (but you'll have to restart the GitPod container and substitute an updated URL for this page to load fully.

**Comment by jywarren** [comment]
So you should be able to just view source on any note page that has an image in its body but no main image. Let's dig deeper to figure out what's going on!!!

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -50,6 +50,9 @@
     <% if @node.main_image %>
         <meta property="og:image" content="https://<%= request.host %><%= @node.main_image.path(:default) %>" />
         <meta name="twitter:image" content="https://<%= request.host %><%= @node.main_image.path(:default) %>">
+        <% elsif @node.images %>
```
Comment (jywarren) [inline app/views/layouts/application.html.erb:53]: Aha so this is why my trick isn't working. If an external image is pasted in, it won't show up in `@node.images`. And, `@node.scraped_image` won't appear here either -- in fact, an image that could have been uploaded and then removed (say with private info) could mistakenly be used if we go with this logic. I'm going to suggest we try swapping all `@node.images` for `@node.scraped_image` here!

```suggestion
        <% elsif scraped_image = @node.scraped_image %>
```

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -50,6 +50,9 @@
...
+        <meta property="og:image" content="https://<%= request.host %><%= @node.images[0] %>" />
```
Comment (jywarren) [inline app/views/layouts/application.html.erb:54]:
```suggestion
        <meta property="og:image" content="<%= scraped_image %>" />
```

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -50,6 +50,9 @@
...
+        <meta name="twitter:image" content="https://<%= request.host %><%= @node.images[0] %>">
```
Comment (jywarren) [inline app/views/layouts/application.html.erb:55]:
```suggestion
        <meta name="twitter:image" content="<%= scraped_image %>">
```

**Comment by jywarren** [comment]
Let's see if this works!!!

**Comment by jywarren** [comment]
Uhhh, darn, i broke everything 🤦

**Comment by jywarren** [comment]
Hmm. Not sure why. Seems unrelated. Will try restarting these!

**Comment by jywarren** [comment]
Yay!!! Gitpod works!


image


Manually checking, it appears in the HTML too:

```
<meta property="og:image" content="https://user-images.githubusercontent.com/24359/90543716-168c0900-e154-11ea-8006-255f06ba85ef.png" />
--
  | <meta name="twitter:image" content="https://user-images.githubusercontent.com/24359/90543716-168c0900-e154-11ea-8006-255f06ba85ef.png">
```

I'm not sure if Facebook is seeing it and I wonder if it's because there are newlines in the description tag, and that's being read first? 

image

**Comment by jywarren** [comment]
Nope! Facebook just takes longer to process the image as it's done asynchronously. It worked there too:

image

OK, now let's figure out why it's failing Travis, and we should be good to resolve this!!! 🎉

**Comment by jywarren** [comment]
Rebasing!

**Comment by jywarren** [comment]
(over https://github.com/publiclab/plots2/pull/8322)!

**Comment by jywarren** [comment]
Nice!!!!

**Comment by Shulammite-Aso** [comment]
> Are you pretty sure it's returning the scraped image from a situation like this? Maybe we need to rebase to get the full ability to post new images inline in new posts?

> Yay!!! Gitpod works!

Happy this works!  I didn't get to confirm it before leaving my desk. Was testing it with twitter but it was just taking time to respond with the card.

**Comment by Shulammite-Aso** [comment]
Trying to check if i'll see this working on gitpod too, but i'm being met with this
Screenshot from 2020-08-18 21-28-08

does this say anything?

**Pull request merged**